### PR TITLE
Rewrite N64 Depth Compare for better performance

### PR DESCRIPTION
**Pull request by loganmc10**
OpenGL extensions exist to ensure fragment shader read/writes don't overlap with each other.

See the performance difference below (tested on Intel+Windows):

Before:

After:

Right now this requires one of 3 extensions: GL_ARB_fragment_shader_interlock , GL_NV_fragment_shader_interlock , GL_INTEL_fragment_shader_ordering. Therefore it leaves out GLES devices, and Mesa, probably some older Nvidia and AMD cards as well. I'm still researching other ways to do this, the most promising is an "atomic spinlock" (https://stackoverflow.com/questions/21538555/broken-glsl-spinlock-glsl-locks-compendium), but I haven't gotten this working yet. Therefore this probably isn't ready to be merged, but I thought I'd leave this here for review.

### CODE REVIEW

**Comment by fzurita** [comment]
Pretty impressive, the performance penalty of n64 depth compare with this is zero from testing so far.

**Comment by AmbientMalice** [comment]
It would be ideal if N64 depth compare could be sped up to the point that enabling it on all necessary games by default would be non-controversial.

**Comment by fzurita** [comment]
One strange issue though, it locks up when multisampling is disabled with the NVidia Shield TV.

**Comment by loganmc10** [comment]
> It would be ideal if N64 depth compare could be sped up to the point that enabling it on all necessary games by default would be non-controversial.

I'm hoping that with improved performance, more people will be inclined to use it and test it, but I don't think it's ready to be the default yet. While I was testing the PR I noticed at least 2 issues that exist in the current master: the background in FIFA 99 is completely missing, and most the Bomb-omb's in Mario Tennis (they stand behind the court in a match) are missing. So N64 Depth Compare still needs some more work in terms of accuracy.

**Comment by gonetz** [comment]
Impressive speed boost!

"N64 Depth Compare still needs some more work in terms of accuracy."

Most likely these are not accuracy issues. There are tricks with frame buffer, which do not take Image depth texture into account.

Also, this mode is not compatible with MSAA. Which lead as to necessity to support post-processing AA. Which leads to general support of post-processing shaders. Which is our bright future, for which I currently has no resources.

**Comment by mudlord** [comment]
@loganmc10

a little nugget of advice, I doubt you will even fucking take it:

https://pastebin.com/tbu5ygym

You are welcome. >_>

**Comment by oddMLan** [comment]
I wish Reshade could get fixed to play nicely with GLideN64

**Comment by loganmc10** [comment]
I added a commit that adds another rendering method, using GL_EXT_shader_framebuffer_fetch. Using this extension, Image Textures aren't used at all, instead the depth data is read/written to framebuffer color attachments.

This extension is supported by Intel on Windows and Linux (starting in Mesa 18.1). It is also supported by PowerVR and Adreno on newer chipsets. I've been testing this with @fzurita , it works properly on Intel and PowerVR, but it doesn't work properly on Adreno...

**Comment by loganmc10** [comment]
So I'm pretty happy with how this PR looks now. This would actually reduce the number of people that could use N64 Depth Compare, since the GPU would need to support Image Textures + one of these extensions, instead of just Image Textures. I think this is the right thing to do though, here is why:

I think it's important that N64 Depth Compare always performs well, right now, we have N64DepthCompare=1 in the INI file for TWINE, just to fix the shadows. So if we allow a fallback to the current (slow) method, a lot of people can't run the game at full speed. It's better to have incorrect shadows and a playable game.

So this PR will try to provide a fast implementation of N64 Depth Compare, otherwise N64 Depth Compare will be disabled. This will allow us to set N64DepthCompare=1 in the INI for games that need it, while still allowing people to play games at full speed if their card doesn't support these extensions.

**Comment by AmbientMalice** [comment]
Would it be worth implementing the z mode compare less hack Glide64 used as a fallback for GPUs that cant handle N64 depth compare? That way games like Toy Story 2 would have correct HUD rendering.

**Comment by loganmc10** [comment]
@fzurita can you try using "NV_pixel_interlock_ordered" instead of "pixel_interlock_ordered". I just realized the NV version exists

**Comment by fzurita** [comment]
Yeah, sure, once I get home tonight.

**Comment by fzurita** [comment]
I get this: 
```
E/GLideN64: shader_compile error: 0(5) : error C3008: unknown layout specifier 'nv_pixel_interlock_ordered'
```

And before you ask, I'm spelling it as ```NV_pixel_interlock_ordered```

**Comment by loganmc10** [comment]
Huh ok weird, I was just reading https://www.khronos.org/registry/OpenGL/extensions/NV/NV_fragment_shader_interlock.txt and I noticed they mentioned both versions in that document, but that's obviously not correct

**Comment by fzurita** [comment]
So apparently the shield TV doesn't support the fragment_interlockNV extension. It's reporting false in "Utils::isExtensionSupported```

**Comment by loganmc10** [comment]
So how is N64 depth compare even working? It should disable it if none of those extensions are enabled

**Comment by fzurita** [comment]
```GL_ARB_fragment_shader_interlock``` does report as being supported.

**Comment by fzurita** [comment]
I did a quick experiment. Commenting out
```m_part += "  if (should_discard) discard;	\n";```
avoids the lockup. Obviously, it creates bad graphic bugs.

**Comment by loganmc10** [comment]
To give some background:GL_ARB_fragment_shader_interlock did not allow me to put the discard statement inside the "block" which is why I created the "should_discard" variable and I do the discard outside the block, perhaps there is a different method of doing it to avoid the crash.

**Comment by fzurita** [comment]
It definitely has to do with ```beginInvocationInterlockARB``` and ```endInvocationInterlockARB```. Commenting those out prevents the lockup.

**Comment by loganmc10** [comment]
Ok yeah so they're both supported, they are essentially the same thing, it's just that only one should be enabled at a time, which is why I added that check

**Comment by fzurita** [comment]
Forcing ```fragment_interlockNV``` to true and ```fragment_interlock``` didn't fix the lockup.

**Comment by olivieryuyu** [comment]
interesting fast depth compare would be great :)

**Comment by gonetz** [comment]
I see no problems with code. I see no regressions - the feature works properly. Performance is really good. Great work!!!

However, locks up issue on NVidia Shield TV must be fixed somehow.
@fzurita "it locks up when multisampling is disabled with the NVidia Shield TV"
N64 depth compare must not be activated at all when MSAA enabled.

**Comment by fzurita** [comment]
Yep, that is true, but that is the only way I could get it to not lock up.

**Comment by fzurita** [comment]
I could force the Android version to always use GLES. There is really no advantage any more to the full OpenGL version with this change. I think this would work with gles because it uses ext_shader_framebuffer_fetch.

**Comment by loganmc10** [comment]
I don't think Nvidia supports ext_shader_framebuffer_fetch, and GL_NV_fragment_shader_interlock should be supported in GLES as well on the Shield. We might just have to exclude the X1 from supporting N64 Depth Compare

**Comment by fzurita** [comment]
That's a pain since it used to work before, at least with the slow method. Maybe we should keep the slow method around for devices that support image textures.

**Comment by loganmc10** [comment]
The problem with keeping the slow method around is you could never have N64 Depth Compare as a default option. It would ruin the performance on some devices. If you only support the fast method then you could turn it on by default, and devices that can't run it fast can just run without it, and still get a pretty good experience.

Is the current/slow method usable on the X1? Do games run at full speed?

**Comment by fzurita** [comment]
It depends on the game, but generally yes. It is around a 50% loss in performance.

**Comment by oddMLan** [comment]
What's the performance difference between no depth compare and the optimized N64 depth compare method?

**Comment by fzurita** [comment]
Depends on the device. In the Nexus player, it was about 10% performance drop. On the Shield tv, the was no performance difference. I'm assuming that on desktop devices, there is no performance penalty.

**Comment by oddMLan** [comment]
Holy moly... that's an astonishing difference! It would make N64 depth compare viable as the default and fix tons of problematic games and little innacuracies with the shadows in many others. However, no MSAA. I guess one could get around it by using a very high internal resolution, at least for now.

**Comment by fzurita** [comment]
The lockup may be fixed by some of the stuff I did in this branch:

https://github.com/fzurita/GLideN64/commits/further_reduce_shader_logic

File: src/Graphics/OpenGLContext/opengl_GLInfo.cpp [hunk 1]
```diff
@@ -79,12 +82,23 @@ void GLInfo::init() {
 		}
 	}
 
+	bool ext_draw_buffers_indexed = isGLESX && (Utils::isExtensionSupported(*this, "GL_EXT_draw_buffers_indexed") || numericVersion >= 32);
 #ifdef EGL
 	if (isGLESX && bufferStorage)
 		g_glBufferStorage = (PFNGLBUFFERSTORAGEPROC) eglGetProcAddress("glBufferStorageEXT");
+	if (isGLESX && numericVerison < 32) {
```
Comment (fzurita) [inline src/Graphics/OpenGLContext/opengl_GLInfo.cpp]: There is a typo: "Verison"

Comment (loganmc10) [reply src/Graphics/OpenGLContext/opengl_GLInfo.cpp]: Whoops, fixed

**Comment by fzurita** [comment]
Including my one commit from the other branch didn't help anything. There is one more extension that can be used that NVidia may support better on the shield TV. It's this: https://www.khronos.org/registry/OpenGL/extensions/KHR/KHR_blend_equation_advanced.txt

**Comment by loganmc10** [comment]
That extension seems widely supported but I'm not sure how it would work for us. It seems to guarantee the coherency of blending operations, but as far as I understand, blending happens after the fragment shader, so it doesn't necessarily guarantee the coherency of fragment shader operations. It's a little unclear to me how you enable coherent behaviour using that extension, it seems to say that it will be enabled by default if it is supported

**Comment by fzurita** [comment]
Yeah, I agree, I realized that after I read more closely into it.

**Comment by fzurita** [comment]
I think I give up at this point I think this can be merged, I'll make a new pull request to remove support from Tegra X1 since it has driver bugs.

**Comment by fzurita** [comment]
Hmmm... I wonder if this can be closed with this: https://github.com/gonetz/GLideN64/issues/692

**Comment by loganmc10** [comment]
Yes I think that would be an example of a game where we could finally just set N64DepthCompare=1 in the INI file for that game and not have to worry about ruining the performance for anyone

**Comment by gonetz** [comment]
>> I think I give up at this point I think this can be merged, I'll make a new pull request to remove support from Tegra X1 since it has driver bugs.

Ok.

**Comment by tony971** [comment]
I'd like to see it on by default for all games unless someone selects MSAA. With no performance impact and a wide array of graphical fixes, it makes sense to just leave it on.

**Comment by loganmc10** [comment]
If anyone wants to try out this new N64 Depth Compare stuff, https://m64p.github.io/ has been updated

To enable N64 Depth Compare you go Settings->Core and Plugin Settings->Video Plugin and look for "EnableN64DepthCompare"

**Comment by oddMLan** [comment]
> I think I give up at this point I think this can be merged, I'll make a new pull request to remove support from Tegra X1 since it has driver bugs.

so no N64 depth compare option for the Nintendo Switch? 😛

**Comment by fzurita** [comment]
Lol, or shield tv, or pixel c

**Comment by loganmc10** [comment]
My phone just got Android 9 and N64 Depth Compare seems to work properly now (Adreno 540)

@fzurita if you have an Android 9 Adreno device you may want to test, I seem to recall this was all messed up on Adreno with Android 8.1

**Comment by fzurita** [comment]
Yep, my phone hasn't been updated. So it really sounds like a driver bug if it's working with Android 9. It is cool too know that once Android 9 comes out for my phone I will have working N64 depth compare though.

**Pull request merged**